### ISSUE TITLE
EF on in-memory data crashes on Left Outer Join

### ISSUE THREAD

**Issue by gregveres** [dotnet/efcore#10673]
I have a query that looks like this:

```C#
from user in db.UserTable
from purchase in db.PurchaseTable.Where(p => p.UserId == user.Id).DefaultIfEmpty()
Where user.Id == 1
select new {
    user = user.Id,
    purchase = (int?)purchase.Id
}.ToListAsync()

```
you can define UserTable and PurchaseTable with just public int Id {get; set;}. 

When this is executed against the database, it creates a fairly efficient query and it executes without exception because purchase will be Null where the user does not have any purchases and in that case, the purchase value of the returned entry will be null. Fantastic. It works and the SQL is straight forward.

But if I then execute this against in-memory data, I get a crash when the in-memory provider is constructing the return value. Even though I can't see the source code in my debugger, I have been able to deduce that it is crashing on accessing Id of purchase because if I only pass in data where every user has a purchase, or if I comment out the purchase = (int?)purchase.Id line, then it works without the crash. 

I am seeing a NullRefereneException. The stack trace is:

```
esult StackTrace:	
at lambda_method(Closure , <>f__AnonymousType150`2 )
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at EntityFramework.Testing.InMemoryDbAsyncEnumerator`1.<MoveNextAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Data.Entity.Infrastructure.IDbAsyncEnumerableExtensions.<ForEachAsync>d__5`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at SquashSpider.Models.CourtBooking.CourtBookingPackageGroup.<FetchMinutesAvailableForUsers>d__26.MoveNext() in C:\Users\gregv\Source\Repos\SquashSpider\SquashSpider\SquashSpider\Models\CourtBooking\CourtBookingPackage.cs:line 98
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at SquashSpider.Tests.Models.CourtBooking.CourtBookingPackageTests.<FetchMinutesAvailableForUsers_Should_return_one_entry_for_player1_When_player1_has_one_purchase_of_30_minutes_with_no_usages>d__11.MoveNext() in C:\Users\gregv\Source\Repos\SquashSpider\SquashSpider\SquashSpider.Tests\Models\CourtBooking\CourtBookingPackageTests.cs:line 103
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
Result Message:	
Test method SquashSpider.Tests.Models.CourtBooking.CourtBookingPackageTests.FetchMinutesAvailableForUsers_Should_return_one_entry_for_player1_When_player1_has_one_purchase_of_30_minutes_with_no_usages threw exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
```

### Steps to reproduce
I will work on putting together a full listing. 
```c#

```

### Further technical details
EF Core version: 6.1.3
Database Provider: EntityFramework.SqlServer v6.1.3 (I am not sure which package provides the in-memory provider)
Operating system: Windows 10
IDE: Visual Studio 2017 15.4

**Comment by maumar** [comment]
looks like EF6 issue

**Comment by gregveres** [comment]
Hi @maumar 
does that mean that you guys aren't considering a fix? Or just that its not a 2.1.0 milestone because it is in EF6, to Core 2?

**Comment by ajcvickers** [comment]
Moved to https://github.com/aspnet/EntityFramework6/issues/450
